### PR TITLE
Version 2

### PR DESCRIPTION
**Pull request by pborges**

### CODE REVIEW

File: smarthome.go [hunk 1]
```diff
@@ -1,118 +1,204 @@
...
+		defer r.Body.Close()
+
+		if agent, ok := s.agents[oauth.GetAgentUserIdFromHeader(r)]; ok {
+			res := s.decodeAndHandle(agent, r.Body)
+			if err := json.NewEncoder(io.MultiWriter(w, log.Writer())).Encode(res); err != nil {
+				log.Println("error encoding sync response:", err)
+				http.Error(w, "error encoding sync response: "+err.Error(), http.StatusInternalServerError)
 			}
+		} else {
+			http.Error(w, "unable to find agentUserId", http.StatusInternalServerError)
```
Comment (pborges) [inline smarthome.go]: This is a bug, SmartHome.Handle() should not return an error, at this point it has "passed" Agent authentication and SmartHome.Handle() needs only return an empty list of devices if it cant find an agent for whatever reason

File: oauth/authprovider.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+type AuthenticationProvider interface {
+	GenerateAuthCode(clientId string, agentUserId string) (string, error)
+	GenerateToken(clientId string, clientSecret string, authCode string) (Token, error)
+	GetAgentUserIdForToken(token string) (agentUserId string, refresh bool, err error)
+}
+
+// a very simple authentication provider
+
+type SimpleAuthenticationProvider struct {
```
Comment (pborges) [inline oauth/authprovider.go]: This struct needs to be re-worked to save its data on disk to survive reboots of the program,

Comment (pborges) [reply oauth/authprovider.go]: This struct should also probably be moved into its own package

Comment (pborges) [reply oauth/authprovider.go]: Or better yet, the concept of "Registered agents" should be abstracted away from the oauth authcodes/tokens etc

File: smarthome.go (hunk 1 above)
Comment (localleon) [reply smarthome.go]: Confirming this bug.

File: oauth/authprovider.go (hunk 2 above)
Comment (localleon) [reply oauth/authprovider.go]: Maybe just serialize the struct as json and write it to disk?

File: cmd/resync/main.go [hunk 3]
```diff
@@ -0,0 +1,14 @@
+package main
+
+import "github.com/pborges/gbridge/homegraph"
+
+func main() {
+	client, err := homegraph.Dial("gbridge-58f4081c3cea.json")
```
Comment (localleon) [inline cmd/resync/main.go]: For a later release, explain on how to get the JWT-Token for the Homegraph API

File: oauth/authprovider.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+type MapBasedAuthProvider struct {
+	db   map[string]mapBasedAuthProviderClient
+	lock sync.Mutex
+}
+
+const charset = "abcdefghijklmnopqrstuvwxyz" +
+	"ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
+
+func GenerateRandomString(length int) string {
+	rand.Seed(time.Now().UnixNano())
```
Comment (localleon) [inline oauth/authprovider.go]: I think it's time to do some real crypto and use "crypto/rand"?

**Review by localleon — COMMENTED** [review]
Overall, i think this is ready to be merged. There are just some bugs and behaviors that need to be worked out.

File: oauth/authprovider.go [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+			return m.agents[agent].Token, nil
+		} else {
+			return Token{}, errors.New("unknown agent")
+		}
+	}
+
+	return Token{}, errors.New("unknown client")
+}
+
+func (m *SimpleAuthenticationProvider) GetAgentUserIdForToken(token string) (agentUserId string, refresh bool, err error) {
```
Comment (localleon) [inline oauth/authprovider.go]: Should be `GetAgentUserIDForToken` , ID capitalized.

File: oauth/authprovider.go (hunk 4 above)
Comment (localleon) [reply oauth/authprovider.go]: If this get's merged, i could probably work on simple stuff like this

Comment (localleon) [reply oauth/authprovider.go]: Fixed this in #7

File: oauth/authprovider.go (hunk 2 above)
Comment (pborges) [reply oauth/authprovider.go]: this was my plan :P I might re-work the struct some before doing so

**Comment by localleon** [comment]
Attributes work!

**Comment by localleon** [comment]
Making good process with the OpenCloseTrait here.

File: oauth/authprovider.go (hunk 2 above)
Comment (localleon) [reply oauth/authprovider.go]: See #10 for this

**Pull request merged**